proposals/0000-markdown-content-redesign.md

```diff
@@ -0,0 +1,126 @@
+- Start Date: 03-10-2022
+- Reference Issues: https://github.com/withastro/rfcs/discussions/5, https://github.com/withastro/rfcs/discussions/118
+- Implementation PR: <!-- leave empty -->
+
+# Summary
+
+A v2.0 API for working with local Markdown files in Astro.
+
+# Example
+
+```astro
+---
+// BASIC USAGE:
+
+// Note: Astro.fetchContent() removed in favor of direct `import.meta.globEager` usage
+// Note: This works in non-Astro framework components!
+const markdownFilesObj = import.meta.globEager('../posts/*.md');
+const markdownFilesArr = Object.values(markdownFilesObj);
+const markdownFilesFiltered = markdownFilesArr.filter(post => post.data.category === 'blog');
+
+// Note: Individual markdown imports are now supported as well.
+const firstPost = await import('../posts/first-post.md');
+---
+<!-- Data is parsed from frontmatter without requiring a render -->
+<h1>{firstPost.data.title}</h1>
+<p>Author: {firstPost.data.author}</p>
+<p><a href={firstPost.url}>Permalink</a></p>
+<!-- Defer rendering as late as possible for performance -->
+<article>{firstPost.getContent()}</article>
+```
+
+
+# Motivation
+
+- **Performance:** Markdown rendering continues to be an expensive part of Astro's runtime. While we can look into faster renderers (ex: Goldmark) there is still a big inefficincy on our end with how we hold the tool. Astro currently renders markdown on import, which means that multi-file markdown imports (via both `import.meta.glob` and `Astro.fetchContent()`) block the response until all imported files are rendered, even if only a subset of these files are ever used on the page after filtering.
+- **Performance (Memory):** For large projects, this also forces Astro to store all rendered Markdown output in memory at the same time, making it difficult to manage memory efficiently. Our community has reported builds maxing out memory limits in a project of only 800 markdown pages, and builds requiring 50+ GB of memory.
+- **Infinite Loops:** By rendering during import, we also introduce a risk for infinite loops when we call `fetchContent()` in a Markdown layout used by a fetched page, bouncing between render and `fetchContent()` infinitely.
+
+
+# Detailed design
+
+- This is implemented in a v2.0 of the internal `'astro:markdown'` Vite plugin
```

Comment:
Would it be possible to actually pull this out of astro core and making it an opt-in integration/plugin?

It may not actually make sense since Astro needs to handle markdown anyway, but requiring a Vite plugin that always claims ownership over every `.md` file import would block any third-party `.md` vite/rollup plugins

Reply:
It would be possible in the new integration system, but I think that would require a really easy-to-use `astro add markdown` CLI command. Otherwise that's a bit too much overhead to add for such an important use-case. 

Also curious to hear others thoughts on this question, it's an interesting one!